lib/src/geometry/polygon.dart

```diff
@@ -0,0 +1,168 @@
+      definition,
+      position: center,
+      size: halfSize * 2,
+      angle: angle,
+    );
+  }
+
+  /// With this constructor you define the [Polygon] from the center of and with
+  /// percentages of the size of the shape.
+  /// Example: [[1.0, 0.0], [0.0, 1.0], [-1.0, 0.0], [0.0, -1.0]]
+  /// This will form a diamond shape within the bounding size box.
+  /// NOTE: Always define your shape is a clockwise fashion
+  Polygon.fromDefinition(
+    this.normalizedVertices, {
+    Vector2 position,
+    Vector2 size,
+    double angle,
+  }) : super(position: position, size: size, angle: angle ?? 0);
+
+  final _cachedScaledShape = ShapeCache<Iterable<Vector2>>();
+
+  /// Gives back the shape vectors multiplied by the size
+  Iterable<Vector2> scaled() {
+    if (!_cachedScaledShape.isCacheValid([size])) {
+      _cachedScaledShape.updateCache(
+        normalizedVertices?.map((p) => p.clone()..multiply(size / 2)),
+        [size.clone()],
+      );
+    }
+    return _cachedScaledShape.value;
+  }
+
+  final _cachedRenderPath = ShapeCache<Path>();
+
+  @override
+  void render(Canvas canvas, Paint paint) {
+    if (!_cachedRenderPath.isCacheValid([position, size])) {
+      _cachedRenderPath.updateCache(
+        Path()
+          ..addPolygon(
+            scaled()
+                .map((point) => (point +
+                        (position + size / 2) +
+                        ((size / 2)..multiply(relativePosition)))
+                    .toOffset())
+                .toList(),
+            true,
+          ),
+        [
+          position.clone(),
+          size.clone(),
+        ],
+      );
+    }
+    canvas.drawPath(_cachedRenderPath.value, paint);
+  }
+
+  final _cachedHitbox = ShapeCache<List<Vector2>>();
+
+  /// Gives back the vertices represented as a list of points which
+  /// are the "corners" of the hitbox rotated with [angle].
+  List<Vector2> hitbox() {
+    // Use cached bounding vertices if state of the component hasn't changed
+    if (!_cachedHitbox.isCacheValid([position, size, angle])) {
+      _cachedHitbox.updateCache(
+        scaled()
+                .map((point) => (point + shapeCenter)
+                  ..rotate(angle, center: anchorPosition))
+                .toList(growable: false) ??
+            [],
+        [shapeCenter, size.clone(), angle],
+      );
+    }
+    return _cachedHitbox.value;
+  }
+
+  /// Checks whether the polygon represented by the list of [Vector2] contains
+  /// the [point].
+  @override
+  bool containsPoint(Vector2 point) {
+    // If the size is 0 then it can't contain any points
+    if (size.x == 0 || size.y == 0) {
+      return false;
+    }
+
+    final vertices = hitbox();
+    for (int i = 0; i < vertices.length; i++) {
+      final edge = getEdge(i, vertices: vertices);
+      final isOutside = (edge.to.x - edge.from.x) * (point.y - edge.from.y) -
+              (point.x - edge.from.x) * (edge.to.y - edge.from.y) >
+          0;
```

Comment:
speaking of the fake cross products this seem to be something like (edge.to - edge.fom) cross (point - edge.from) > 0 (or the other way around)

Reply:
we can think of it later